src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs

```diff
@@ -0,0 +1,102 @@
+                }
+            }
+
+            return new HttpPipeline(transport, currentPolicy);
+        }
+
+        public static HttpPipelineOptions AddLogging(this HttpPipelineOptions options, int[] excludeCodes = null)
+        {
+            return options.Append(HttpPipelineSection.PostRetry, next => new LoggingPolicy(next, excludeCodes));
+        }
+
+        public static HttpPipelineOptions AddFixedRetry(this HttpPipelineOptions options, int maxRetries, TimeSpan delay, params int[] retriableCodes)
+        {
+            return options.Replace(HttpPipelineSection.Retry, next => new FixedPolicy(next, maxRetries, delay, retriableCodes));
+        }
+
+        public static HttpPipelineOptions AddHeaders(this HttpPipelineOptions options, params HttpHeader[] headers)
+        {
+            return options.Replace(HttpPipelineSection.PreRetry, next => new AddHeadersPolicy(next, headers));
+        }
+
+        public static HttpPipelineOptions UseHttpClient(this HttpPipelineOptions options, HttpClient client)
+        {
+            options.TransportPolicy =  () => {
+                return new HttpClientTransport()
+                {
+                    Client = client
+                };
+            };
+            return options;
+        }
+
+        public static HttpPipelineOptions UseHttpClient(this HttpPipelineOptions options, Action<HttpClientHandler> configure)
+        {
+            options.TransportPolicy =  () => {
+                var handler = new HttpClientHandler();
+                configure(handler);
+                return new HttpClientTransport()
+                {
+                    Client = new HttpClient(handler)
+                };
+            };
+
+            return options;
+        }
+
+        public static HttpPipelineOptions Append(this HttpPipelineOptions options, HttpPipelineSection section, Func<HttpPipelinePolicy, HttpPipelinePolicy> policyFactory)
+        {
+            options.PipelinePolicies.Add(new PolicyRegistration(section, policyFactory));
+            return options;
+        }
+
+        public static HttpPipelineOptions TryAppend(this HttpPipelineOptions options, HttpPipelineSection section, Func<HttpPipelinePolicy, HttpPipelinePolicy> policyFactory)
+        {
+            if (options.PipelinePolicies.All(p => p.Section != section))
+            {
+                options.PipelinePolicies.Add(new PolicyRegistration(section, policyFactory));
+            }
+            return options;
+        }
+
+        public static HttpPipelineOptions Replace(this HttpPipelineOptions options, HttpPipelineSection section, Func<HttpPipelinePolicy, HttpPipelinePolicy> policyFactory)
+        {
+            foreach (var registration in options.PipelinePolicies.Where(registration => registration.Section == section).ToArray())
+            {
+                options.PipelinePolicies.Remove(registration);
+            }
+
+            return options.Append(section, policyFactory);
+        }
```

Comment:
We need the ability to add an arbitrary transport. Also, we need the ability to remove policies added by the service, eg. logging policy. Also, it would be nice if the telemetry policy was something that could be removed.

Reply:
We need to decide who can override who.

It's either "customer can override the client default" or "client can add default policies to what client passed in" we can't have both, so which one should we pick?

Reply:
Client suggests defaults, user can tweak them, client can veto some tweaks to ensure it works properly. For example, client needs to suggest retry policy, the user can tweak it's settings, but if the user removes authentication policy, the client will make sure that despite of this, the policy is present in the pipeline.

Reply:
What if the user wants to use a completely different retry policy and ignore the client provided one?